### PR TITLE
DAOS-5465 csum: Corrupt BIO Address through VOS

### PR DESCRIPTION
**Pull request by ryon-jensen**
Add the ability to set a bio address as corrupted using the VOS
iter API. This will be necessary for the checksum scrubber
to be able to mark data as corrupted when silent data
corruption is detected.

- Use extra padding in the bio_addr_t structure for flags and add a
  corrupted flag to indicate that data is corrupted.
- Add corrupted field to necessary structs in VOS to facilitate
  the setting/getting of the corrupted status of the data.
- Add vos_iter_corrupt api for setting the current record
  as corrupted.

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

### CODE REVIEW

**Comment by ryon-jensen** [comment]
This is a different approach to PR5444

**Review by jolivier23 — DISMISSED** [review]

File: src/include/daos_srv/bio.h [hunk 1]
```diff
@@ -28,7 +35,7 @@ typedef struct {
 	/* Is the address a hole ? */
 	uint16_t	ba_hole;
 	uint16_t	ba_dedup;
-	uint16_t	ba_padding;
+	uint16_t	ba_flags;
```
[the remark this answers is not in the document]
Comment (jolivier23) [reply src/include/daos_srv/bio.h:38]: This should probably go in a separate change if we do it.  One that we might want to push to 1.2 along with space reservation for a btree in the pool

[the remark this answers is not in the document]
Comment (ryon-jensen) [reply src/include/daos_srv/bio.h:38]: Makes sense. I can split it in to a different PR and create a separate ticket for it if it indeed needs to land to 1.2. I'm not in the meetings where 1.2 stuff is discussed, so let me know what's decided.

File: src/include/daos_srv/vos.h [hunk 2]
```diff
@@ -916,6 +916,18 @@ vos_iter_copy(daos_handle_t ih, vos_iter_entry_t *entry,
...
+ * Mark the current record at the current cursor as being corrupt. This function
+ * will call into the btree/evtree for them to update the bio_addr with the
+ * appropriate flag to indicate the bio_addr references corrupt data.
+ *
+ * \param ih	[IN]	Iterator open handle.
+ *
+ * \return		Zero on Success
+ */
+int
+vos_iter_corrupt(daos_handle_t ih);
```
[the remark this answers is not in the document]
Comment (ryon-jensen) [reply src/include/daos_srv/vos.h:929]: Ya, I agree. vos_iter_corrupt is too specialized. :)

File: src/vos/evt_iter.c [hunk 3]
```diff
@@ -550,6 +550,33 @@ int evt_iter_delete(daos_handle_t ih, struct evt_entry *ent)
...
+	iter = &tcx->tc_iter;
+	rc = evt_iter_is_ready(iter);
+	if (rc != 0)
+		return rc;
+
+	trace = &tcx->tc_trace[tcx->tc_depth - 1];
+	desc = evt_node_desc_at(tcx, evt_off2node(tcx, trace->tr_node),
+				trace->tr_at);
+	D_DEBUG(DB_IO, "Setting record bio_addr flag to corrupted\n");
+	BIO_ADDR_SET_CORRUPTED(&desc->dc_ex_addr);
```
Comment (NiuYawei) [inline src/vos/evt_iter.c:575]: I'm wondering why don't make it into a transaction to ensure the persistency (like what you did for dbtree_iter_corrupt()).

File: src/vos/vos_tree.c [hunk 4]
```diff
@@ -749,6 +748,18 @@ svt_rec_update(struct btr_instance *tins, struct btr_record *rec,
...
 }
 
+static int
+svt_rec_corrupt(struct btr_instance *tins, struct btr_record *rec)
+{
+	struct vos_irec_df	*irec;
+
+	irec = vos_rec2irec(tins, rec);
+	D_DEBUG(DB_IO, "Setting record bio_addr flag to corrupted\n");
+	BIO_ADDR_SET_CORRUPTED(&irec->ir_ex_addr);
```
Comment (NiuYawei) [inline src/vos/vos_tree.c:758]: You did a in-place update (within PMDK transaction) without making snapshot by umem_tx_add().

**Review by NiuYawei — CHANGES_REQUESTED** [review]

File: src/include/daos_srv/bio.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryon-jensen) [reply src/include/daos_srv/bio.h:38]: I created PR #5510 for the bio_addr_t changes.

**Comment by ryon-jensen** [comment]
Thanks @NiuYawei for your feedback. I'm not as familiar with the umem transaction stuff. Would you take a look at my updates in response to your feedback? There's still more updates that I need to make, but I want to make sure I've got the umem stuff right.

File: src/vos/evt_iter.c [hunk 5]
```diff
@@ -571,10 +571,16 @@ evt_iter_corrupt(daos_handle_t ih)
 	trace = &tcx->tc_trace[tcx->tc_depth - 1];
 	desc = evt_node_desc_at(tcx, evt_off2node(tcx, trace->tr_node),
 				trace->tr_at);
+	rc = evt_tx_begin(tcx);
+	if (rc != DER_SUCCESS)
+		return rc;
+
+	rc = umem_tx_add(&tcx->tc_umm, trace->tr_node, sizeof(*desc));
```
Comment (NiuYawei) [inline src/vos/evt_iter.c:578]: I think we don't have to making snapshot for the whole desc,  adding 'dc_tx_addr' should be sufficient.

File: src/vos/vos_tree.c [hunk 6]
```diff
@@ -752,8 +752,14 @@ static int
 svt_rec_corrupt(struct btr_instance *tins, struct btr_record *rec)
 {
 	struct vos_irec_df	*irec;
+	int			 rc;
 
 	irec = vos_rec2irec(tins, rec);
+
+	rc = umem_tx_add(&tins->ti_umm, rec->rec_off, sizeof(*irec));
```
Comment (NiuYawei) [inline src/vos/vos_tree.c:759]: Not a big deal, adding 'ir_ex_addr' only?

**Review by NiuYawei — APPROVED** [review]

File: src/vos/evt_iter.c [hunk 7]
```diff
@@ -550,6 +550,39 @@ int evt_iter_delete(daos_handle_t ih, struct evt_entry *ent)
...
+	trace = &tcx->tc_trace[tcx->tc_depth - 1];
+	desc = evt_node_desc_at(tcx, evt_off2node(tcx, trace->tr_node),
+				trace->tr_at);
+	rc = evt_tx_begin(tcx);
+	if (rc != DER_SUCCESS)
+		return rc;
+
+	rc = umem_tx_add(&tcx->tc_umm, trace->tr_node, sizeof(*desc));
+	D_DEBUG(DB_IO, "Setting record bio_addr flag to corrupted\n");
+	BIO_ADDR_SET_CORRUPTED(&desc->dc_ex_addr);
```
Comment (Nasf-Fan) [inline src/vos/evt_iter.c:580]: Do that when "if (rc == 0)".

**Pull request closed without merging**